plugins/velocity/src/main/java/org/apache/struts2/views/velocity/StrutsVelocityContext.java

```diff
@@ -96,11 +96,15 @@ protected Object chainedContextGet(String key) {
             return null;
         }
         for (VelocityContext chainedContext : chainedContexts) {
-            Object val = chainedContext.internalGet(key);
+            Object val = chainedContext.get(key);
             if (val != null) {
                 return val;
             }
         }
         return null;
     }
+
+    public ValueStack getValueStack() {
+        return stack;
+    }
```

Comment:
Do we need to have this public? Also this requires to cast to `StrutsVelocityContext`, wouldn't be better to have a marking interface `ValueStackAware` or `ValueStackProvider` to expose the ValueStack?

Reply:
I think it has to be but I wonder if Directives can simply obtain the ValueStack from the ActionContext? It's not clear to me if the ValueStack on the ActionContext changes between the time of Velocity context creation and directive rendering. I recall this was how the ValueStack was obtained in WebWork 2.1 but I presume it was changed for a reason?

But also, there's no change in terms of security as the stack was already exposed on the `StrutsVelocityContext` instance using `internalGet("stack")` or `get("stack")`.

And yep I can definitely use a marker interface to allow more flexibility in the Velocity context implementation used by applications.

Reply:
It shouldn't change as far I understand, yet I'm not sure if this is true :) Let's leave it as is, at some point I will solve this puzzle ;-)